utils/fixed_decimal/src/decimal.rs

```diff
@@ -618,6 +618,92 @@ impl FixedDecimal {
         self.check_invariants();
     }
 
+    /// Concatenate another `FixedDecimal` into this `FixedDecimal`.
+    ///
+    /// The two decimals must represent non-overlapping ranges of magnitudes. If the range of
+    /// nonzero digits overlaps between `self` and `other`, an `Err` is returned.
+    ///
+    /// The magnitude range of `self` will be increased if `other` covers a larger range.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use fixed_decimal::FixedDecimal;
+    ///
+    /// let mut integer = FixedDecimal::from(123);
+    /// let fraction = FixedDecimal::from(456)
+    ///     .multiplied_pow10(-3)
+    ///     .expect("in-range");
+    ///
+    /// assert_eq!("123.456", integer.concatenated(fraction).expect("nonoverlapping").to_string());
+    /// ```
+    pub fn concatenated(mut self, other: FixedDecimal) -> Result<Self, Error> {
+        match self.concatenate(other) {
+            Ok(()) => Ok(self),
+            Err(err) => Err(err),
+        }
+    }
+
+    /// Concatenate another `FixedDecimal` into this `FixedDecimal`.
+    ///
+    /// The two decimals must represent non-overlapping ranges of magnitudes. If the range of
+    /// nonzero digits overlaps between `self` and `other`, an `Err` is returned.
+    ///
+    /// The magnitude range of `self` will be increased if `other` covers a larger range.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use fixed_decimal::FixedDecimal;
+    ///
+    /// let mut integer = FixedDecimal::from(123);
+    /// let fraction = FixedDecimal::from(456)
+    ///     .multiplied_pow10(-3)
+    ///     .expect("nonoverlapping");
+    ///
+    /// integer.concatenate(fraction);
+    ///
+    /// assert_eq!("123.456", integer.to_string());
+    /// ```
+    pub fn concatenate(&mut self, other: FixedDecimal) -> Result<(), Error> {
+        let self_left = self.nonzero_magnitude_left();
+        let self_right = self.nonzero_magnitude_right();
+        let other_left = other.nonzero_magnitude_left();
+        let other_right = other.nonzero_magnitude_right();
+        if self.is_zero() {
+            // Operation will succeed. We can move the digits into self.
+            self.digits = other.digits;
+            self.magnitude = other.magnitude;
+        } else if other.is_zero() {
+            // No changes to the digits are necessary.
+        } else if self_left >= other_right && self_right <= other_left {
```

Comment:
issue: this onyl handles the case where other is contained within self, not the other cases of self contained in other, or partial overlap on either side

I think we need to ensure that self's l/r are either both greater than or both less than other's l/r.

This can be done via `self_left < other_right || other_left < self_right` i think

Reply:
Note that `self_left >= self_right` and `other_left >= other_right` because "left magnitudes" are greater than "right magnitudes".

I *think* that what I wrote is the correct way to check if two inclusive ranges are overlapping.

https://stackoverflow.com/a/3269471/1407170

Reply:
Yeah, I incorporated that into my formula: it's got to be an or condition because there are two potential configurations when it's non-overlapping, one where self is entirely greater than other and one where it's entirely less

Reply:
The condition is true when the intervals are overlapping. I think your expression is the exact opposite of mine.

I wrote:

```
self_left >= other_right && self_right <= other_left
```

You wrote:

```
self_left < other_right || other_left < self_right
```

I think those two expressions are `!` of each other.

Reply:
oh, i see, i wrote the condition for "there are no overlaps"